cdap-examples/Purchase/src/main/java/co/cask/cdap/examples/purchase/PurchaseHistoryBuilder.java

```diff
@@ -134,8 +134,9 @@ public void reduce(Text customer, Iterable<Text> values, Context context)
 
       PurchaseHistory purchases = new PurchaseHistory(customer.toString(), userProfile);
       int numPurchases = 0;
-      for (Text val : values) {
-        purchases.add(new Gson().fromJson(val.toString(), Purchase.class));
+      for (Purchase val : values) {
+        purchases.add(new Purchase(val.getCustomer(), val.getProduct(), val.getQuantity(), val.getPrice(),
```

Comment:
The reducer does not store all values in memory. While looping through the Iterable value list, each Object instance is re-used, so it only keeps one instance around at a given time. Because of this we need to make the copy of the object - http://stackoverflow.com/questions/11024638/hadoop-reducer-values-in-memory